prio_processor/spark/commands.py

```diff
@@ -0,0 +1,355 @@
+@click.option(
+    "--scale", type=int, default=1, help="Factor to duplicate shares for output."
+)
+@click.option("--partition-size-mb", type=int, default=250, help="")
+def generate(
+    batch_id,
+    n_data,
+    public_key_hex_internal,
+    public_key_hex_external,
+    output_a,
+    output_b,
+    n_rows,
+    scale,
+    partition_size_mb,
+):
+    shares = (
+        spark_session()
+        .range(n_rows * n_data)
+        .select(
+            (F.col("id") % n_rows).alias("row_id"),
+            F.when(F.rand() > 0.5, 1).otherwise(0).alias("payload"),
+        )
+        .groupBy("row_id")
+        .agg(F.collect_list("payload").alias("payload"))
+        .select(
+            F.pandas_udf(
+                partial(
+                    udf.encode,
+                    batch_id,
+                    n_data,
+                    public_key_hex_internal,
+                    public_key_hex_external,
+                ),
+                returnType="a: binary, b: binary",
+            )("payload").alias("shares")
+        )
+        # repeat this data `scale` times
+        .withColumn("_repeat", F.explode(F.array_repeat(F.lit(0), scale)))
+        .drop("_repeat")
+        .withColumn("id", F.udf(lambda: str(uuid4()), returnType="string")())
+    )
+    shares.cache()
+    # we can make an estimate with just a single row, since the configuration
+    # is the same here.
+    row = shares.first()
+    num_partitions = math.ceil(
+        (len(b64encode(row.shares.a)) + len(str(uuid4())))
+        * n_rows
+        * scale
+        * 1.0
+        / 10 ** 6
+        / partition_size_mb
+    )
+    click.echo(f"writing {num_partitions} partitions")
+
+    # caching is required for a stable id, unless we take advantage of partitioned writing
+    repartitioned = shares.repartitionByRange(num_partitions, "id").cache()
+    repartitioned.select("id", F.base64("shares.a").alias("payload")).write.json(
+        output_a, mode="overwrite"
+    )
+    repartitioned.select("id", F.base64("shares.b").alias("payload")).write.json(
+        output_b, mode="overwrite"
+    )
+
+
+@entry_point.command()
+@data_config
+@public_key
+@input_1
+@output_2
+def encode_shares(
+    batch_id,
+    n_data,
+    public_key_hex_internal,
+    public_key_hex_external,
+    input,
+    output_a,
+    output_b,
+):
+    click.echo("Running encode shares")
+    spark = spark_session()
+    shares = (
+        spark.read.json(input)
+        .withColumn("pid", F.spark_partition_id())
+        .groupBy("pid")
+        .applyInPandas(
+            lambda pdf: udf.encode(
+                batch_id, n_data, public_key_hex_internal, public_key_hex_external, pdf
+            ),
+            schema="a: binary, b: binary",
+        )
+        .withColumn("id", F.udf(lambda: str(uuid4()), returnType="string")())
+    )
+    shares.cache()
+
+    # should this guarantee valid ranges for each partition?
```

Comment:
This comment is a bit concerning. I assume the answer is yes? 😃

Reply:
Yeah, but it was less clear before it was using the range partitioning.